File: 2024/day2/solution.py
def is_safe(levels):
    increasing = levels[0] < levels[1]
    safe = True
    for i in range(len(levels) - 1):
        diff = levels[i + 1] - levels[i] if increasing else levels[i] - levels[i + 1]
        if diff not in diffs:
            safe = False
            break
    return safe

with open("puzzle_input.txt", "r") as puzzle:
    level_reports = puzzle.read().splitlines()
    diffs = (1, 2, 3)
    acc = 0
    for report in level_reports:
        levels = list(map(lambda x: int(x), report.split(" ")))
        increasing = levels[0] < levels[1]
        safe = is_safe(levels)
        if safe: acc += 1
    print("Part 1:", acc)

    acc = 0
    for report in level_reports:
        levels = list(map(lambda x: int(x), report.split(" ")))
        safe = is_safe(levels)
        if safe: acc += 1
        else:
            safe = False
            for i in range(len(levels)):
                new_levels = [levels[j] for j in range(len(levels)) if j != i]
                safe = is_safe(new_levels)
                if safe:
                    break
            if safe: acc += 1
    print("Part 2:", acc)
            

File: 2024/day2/solution.rs
use std::fs;

fn is_safe(levels: &Vec<i32>, diffs: &Vec<i32>) -> bool {
    let increasing = levels[0] < levels[1];
    let mut safe = true;
    for i in 0..(levels.len() - 1) {
        let diff = if increasing { levels[i + 1] - levels[i] } else { levels[i] - levels[i + 1] };
        if !diffs.contains(&diff) {
            safe = false;
            break;
        }
    }
    return safe
}

fn main() {
    let file_path = "puzzle_input.txt";
    let puzzle = fs::read_to_string(file_path).expect("Couldn't read file.");
    let reports = puzzle.lines();
    let diffs = vec![1, 2, 3];

    // Part 1
    let mut acc = 0;
    for report in reports.clone() {
        let levels: Vec<i32> = report.split_whitespace().into_iter().map(|lvl| { lvl.parse::<i32>().unwrap() }).collect();
        if is_safe(&levels, &diffs) {
            acc += 1;
        }
    }
    println!("Part 1: {}", acc);

    // Part 2
    acc = 0;
    for report in reports.clone() {
        let levels: Vec<i32> = report.split_whitespace().into_iter().map(|lvl| { lvl.parse::<i32>().unwrap() }).collect();
        if is_safe(&levels, &diffs) {
            acc += 1;
        } else {
            for i in 0..levels.len() {
                let mut new_levels = levels.clone();
                new_levels.remove(i);
                let safe = is_safe(&new_levels, &diffs);
                if safe {
                    acc += 1;
                    break;
                }
            }
        }
    }
    println!("Part 2: {}", acc);
}